### PR TITLE
feat: roundabout 

### PR DESCRIPTION
**Pull request by vasco-santos**
Adds roundabout :arrows_counterclockwise: stack to w3infra. It includes two HTTP Endpoints:
- `GET /{carCid}` - Gets the content of a CAR file by its CID. Behind the scenes, this request is redirected for a presigned URL created on the spot
- `HEAD /{carCid}` - Gets header with `location` with presigned URL to download CAR file
  - Comes with a 3 days default expiration date (default chose given is the actually limit imposed for a contract to be sealed in Spade), but can be requested to be bigger via query param

This PR was added in this repo under the following decisions:
- we do not want to create more and more repos with SST projects, specially for small context APIs
- we do not want to add this feature within the scope of `up.web3.storage` and therefore a new API should be created

As drawback it is worth pointing out that creating more and more Stacks can also bite us. There are account limits in the number of Cloudformation Stacks and Roles we can have (we can also ask AWS to grant us more).

CF docs for presigned URL https://developers.cloudflare.com/r2/examples/aws/aws-sdk-js-v3/#generate-presigned-urls

### CODE REVIEW

**Comment by vasco-santos** [comment]
See how to use:

```console
$ curl https://pr205.roundabout.web3.storage/presigned-url/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq
https://carpark-staging-0.fffa4b4363a7e5250af8357087263b3a.r2.cloudflarestorage.com/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq.car?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=a314d2872c5c092e911e3e2c7b5e5c3f%2F20230614%2Fauto%2Fs3%2Faws4_request&X-Amz-Date=20230614T132408Z&X-Amz-Expires=259200&X-Amz-Signature=78f593c61d3babfa66af298e7be0cce9bca52828ab2c34232add46e06a60f833&X-Amz-SignedHeaders=host&x-id=GetObject%

$ curl https://pr205.roundabout.web3.storage/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq
{"status":302,"headers":{"location":[{"key":"Location","value":"https://carpark-staging-0.fffa4b4363a7e5250af8357087263b3a.r2.cloudflarestorage.com/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq.car?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=a314d2872c5c092e911e3e2c7b5e5c3f%2F20230614%2Fauto%2Fs3%2Faws4_request&X-Amz-Date=20230614T132534Z&X-Amz-Expires=900&X-Amz-Signature=03f996df9fc4fd23f529f4c46227d7f12e8eacd040c924bcaaef4c8c34ec3d41&X-Amz-SignedHeaders=host&x-id=GetObject"}]}}%
```

**Comment by vasco-santos** [comment]
@alanshaw suggested that perhaps HTTP method should be a POST. By HTTP semantics POST would be used when creating a resource. I see this more like a Map where you give a Key and service gives back a value (pre-signed-url) that you can use. There is no state, and therefore using GET. 

Thoughts?

**Comment by alanshaw** [comment]
> @alanshaw suggested that perhaps HTTP method should be a POST. By HTTP semantics POST would be used when creating a resource. I see this more like a Map where you give a Key and service gives back a value (pre-signed-url) that you can use. There is no state, and therefore using GET.
> 
> Thoughts?

...understood, but the presigned URL does not already exist, you have to do some work to create it.

**Comment by alanshaw** [comment]
> > @alanshaw suggested that perhaps HTTP method should be a POST. By HTTP semantics POST would be used when creating a resource. I see this more like a Map where you give a Key and service gives back a value (pre-signed-url) that you can use. There is no state, and therefore using GET.
> > Thoughts?
> 
> ...understood, but the presigned URL does not already exist, you have to do some work to create it.

Not strong feelings given `GET /{carCid}` creates a resource and then redirects to it...

File: roundabout/functions/presigned-url.js [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  const {
+    BUCKET_ENDPOINT,
+    BUCKET_REGION,
+    BUCKET_ACCESS_KEY_ID,
+    BUCKET_SECRET_ACCESS_KEY,
+    BUCKET_BUCKET_NAME,
+  } = getEnv()
+
+  const expiresIn = request.queryStringParameters?.expires ?
+    Number(request.pathParameters?.expires)
```
Comment (alanshaw) [inline roundabout/functions/presigned-url.js:30]:
```suggestion
    parseInt(request.queryStringParameters?.expires)
```

Comment (alanshaw) [reply roundabout/functions/presigned-url.js:30]: Needs a test ;)

File: roundabout/functions/presigned-url.js [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @param {import('aws-lambda').APIGatewayProxyEventV2} request
+ */
+export async function presignedUrlGet(request) {
+  const {
+    BUCKET_ENDPOINT,
+    BUCKET_REGION,
+    BUCKET_ACCESS_KEY_ID,
+    BUCKET_SECRET_ACCESS_KEY,
+    BUCKET_BUCKET_NAME,
```
Comment (alanshaw) [inline roundabout/functions/presigned-url.js:26]:
```suggestion
    BUCKET_NAME,
```

File: docs/roundabout.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+### `GET /{carCid}`
+
+Gets the content of a CAR file (by its CID) from a signed URL directly to the bucket where it is being stored. Behind the scenes, this request will return a `302 Redirect` to a created signed URL.
+
+```console
+curl -L -v https://roundabout.web3.storage/presigned-url/bagbaiera222226db4v4oli5fldqghzgbv5rqv3n4ykyfxk7shfr42bfnqwua --output bagbaiera222226db4v4oli5fldqghzgbv5rqv3n4ykyfxk7shfr42bfnqwua.car
+```
+
+### `GET /presigned-url/{carCid}`
```
Comment (alanshaw) [inline docs/roundabout.md:17]: Do we need this endpoint, could we not just send a HEAD request like `HEAD /${carCid}` and extract the `Location` header?

**Review by alanshaw — CHANGES_REQUESTED** [review]

**Comment by alanshaw** [comment]
Maybe we should just host it in CF? :trollface:

File: docs/roundabout.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+Gets a presigned URL to get the content of a CAR file (by its CID). This presigned URL is valid by 3 days by default. This also supports a query parameter `expires` with the number of seconds this presigned URL should be valid for.
```
[the remark this answers is not in the document]
Comment (vasco-santos) [reply docs/roundabout.md]: 1 week and now documented

File: roundabout/functions/presigned-url.js (hunk 1 above)
Comment (vasco-santos) [reply roundabout/functions/presigned-url.js]: refactored for a util so that we can test

**Comment by vasco-santos** [comment]
New results

```console
$ curl --head https://pr205.roundabout.web3.storage/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq
HTTP/2 302
date: Wed, 21 Jun 2023 10:34:15 GMT
content-length: 0
location: https://carpark-staging-0.fffa4b4363a7e5250af8357087263b3a.r2.cloudflarestorage.com/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq.car?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=a314d2872c5c092e911e3e2c7b5e5c3f%2F20230621%2Fauto%2Fs3%2Faws4_request&X-Amz-Date=20230621T103415Z&X-Amz-Expires=259200&X-Amz-Signature=c88c38cb8b17f8d90d1a6c4b181de5385caaa430636bd5926485eb3105420ebd&X-Amz-SignedHeaders=host&x-id=GetObject
apigw-requestid: G3XGGjqFiYcEM9w=

$ curl -L https://pr205.roundabout.web3.storage/bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq --output bagbaiera223xmiutg62dsthdyd6kqgsft25knslnlaxxvwe6nc4zrwezezeq.car
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   225  100   225    0     0    193      0  0:00:01  0:00:01 --:--:--  4326
```

**Comment by vasco-santos** [comment]
HEAD request was a great suggestion, thanks! 💯

**Review by alanshaw — APPROVED** [review]

**Pull request merged**